### PR TITLE
Forward measurement type to c8y

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

Converter must parse the `measurement type` from topic if its present and use it in the converted `c8y` payload.
The priority is as below
Use `measurement type` from `payload` if it is present
Otherwise, use the `measurement type` that is mentioned in the topic.
If both are not present then use the default to `ThinEdgeMeasurement`.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://github.com/thin-edge/thin-edge.io/issues/2215

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/extensions/c8y_mapper_ext/src/json.rs [hunk 1]
```diff
@@ -74,7 +76,7 @@ mod tests {
 
         let entity = EntityMetadata::main_device("foo".to_string());
         let output =
-            from_thin_edge_json_with_timestamp(single_value_thin_edge_json, timestamp, &entity);
+            from_thin_edge_json_with_timestamp(single_value_thin_edge_json, timestamp, &entity, "");
```
Comment (PradeepKiruvale) [inline crates/extensions/c8y_mapper_ext/src/json.rs:79]: Not that happy to use an empty string. I feel better to use an `Option` here.

File: crates/extensions/c8y_mapper_ext/src/json.rs [hunk 2]
```diff
@@ -36,18 +36,20 @@ pub enum CumulocityJsonError {
...
 }
 
 fn from_thin_edge_json_with_timestamp(
     input: &str,
     timestamp: OffsetDateTime,
     entity: &EntityMetadata,
+    m_type: &str,
 ) -> Result<String, CumulocityJsonError> {
-    let mut serializer = serializer::C8yJsonSerializer::new(timestamp, entity);
+    let mut serializer = serializer::C8yJsonSerializer::new(timestamp, entity).with_type(m_type);
```
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/json.rs:52]: I would simply add an extra argument to the constructor as this is the only place where a `C8yJsonSerializer` is built.

```suggestion
    let mut serializer = serializer::C8yJsonSerializer::new(timestamp, entity, m_type);
```

File: crates/extensions/c8y_mapper_ext/src/json.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/json.rs:79]: In this specific case, `Option<String>` makes things a bit artificial.

- The c8y mapper will have to deal with `None` and `Some("")` as a non empty type string is mandatory.
- The measurement type received on the topic is actually a string, possibly empty.

Comment (PradeepKiruvale) [reply crates/extensions/c8y_mapper_ext/src/json.rs:79]: I agree, I will keep this as `string`.

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 3]
```diff
@@ -1433,6 +1437,123 @@ mod tests {
...
+            .collect();
+        assert_eq!(
+            out_first_messages,
+            vec![
+                expected_smart_rest_message,
+                expected_c8y_json_message.clone()
+            ]
+        );
+    }
+
```
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:1556]: Better to add a test for the empty measurement type case: `te/device/main///m/` as well.

File: tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot [hunk 4]
```diff
@@ -19,6 +19,16 @@ Thin-edge devices support sending simple measurements with custom type
...
     Log    ${measurements}    
 
+Thin-edge devices support sending simple measurements with custom type in topic
+    Execute Command    tedge mqtt pub te/device/main///m/CustomType_topic '{ "temperature": 25 }'
+    ${measurements}=    Device Should Have Measurements    minimum=1    maximum=1    type=CustomType_topic    value=temperature    series=temperature
+    Log    ${measurements}
+
+
+Thin-edge devices support sending simple measurements with custom type in payload
+    Execute Command    tedge mqtt pub te/device/main///m/CustomType_payload '{ "type":"CustomType_payload","temperature": 25 }'
```
Comment (albinsuresh) [inline tests/RobotFramework/tests/cumulocity/telemetry/thin-edge_device_telemetry.robot:29]:
```suggestion
    Execute Command    tedge mqtt pub te/device/main///m/CustomType_topic '{ "type":"CustomType_payload","temperature": 25 }'
```
Just to ensure that the `type` is picked up from the payload itself.

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 3 above)
Comment (PradeepKiruvale) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:1556]: This is already covered by `test_convert_small_measurement`

**Review by didier-wenzek — APPROVED** [review]
Approved. Thank you.

**Comment by didier-wenzek** [comment]
> LGTM. However, now @reubenmiller suggested to use `.type` for the field name [here](https://github.com/thin-edge/thin-edge.io/issues/2215#issuecomment-1706351798), so I would wait until it gets more clarified.

I don't get your point @rina23q. This PR is already using the `type` field of the message, if any. This is in fact inherited from the [current behavior](https://github.com/thin-edge/thin-edge.io/issues/770#issuecomment-1516315634)

**Comment by rina23q** [comment]
> I don't get your point @rina23q. This PR is already using the type field of the message, if any. This is in fact inherited from the https://github.com/thin-edge/thin-edge.io/issues/770#issuecomment-1516315634

I really got confused in @reubenmiller's comment. I know `type` is exactly the current behaviour, but he wrote `.type`, not `type`. What is supposed to be?

**Comment by reubenmiller** [comment]
> > I don't get your point @rina23q. This PR is already using the type field of the message, if any. This is in fact inherited from the https://github.com/thin-edge/thin-edge.io/issues/770#issuecomment-1516315634
> 
> 
> 
> I really got confused in @reubenmiller's comment. I know `type` is exactly the current behaviour, but he wrote `.type`, not `type`. What is supposed to be?

It is the "type" property. .type is just a common way to say it is part of an object.

**Review by reubenmiller — APPROVED** [review]
Approved
